묵찌빠 게임 [Step 2]Charles, Toy

Developer-Groo opened the pull request:
@mint3382  안녕하세요 민트! step2 Pr보내 드립니다~

<h3>고민했던 점</h3>
<ul>
<li>컴퓨터의 턴과 사용자의 턴을 구분하는 방법을 고민했습니다.</li>
<div> -isUserTurn변수를 사용해서 컴퓨터의 턴과 사용자의 턴을 구분하여 담을 수 있도록 해결했습니다.</div>
<div> -checkPlayerTurn함수를 사용해서 isUserTurn변수의 값에 따라 턴과 메뉴를 출력하도록 구현했습니다.</div>
</ul>

<h3>조언을 얻고 싶은 점</h3>
<ul>
<li>case를 나누는 단위</li>
<div> -가위,바위,보와 묵,찌,빠만 case를 나눠서 구현을 했는데 다른 팀의 코드를 보니 승패여부 등 세세하게 case를 나눠서 구현을 한 것 을 확인했습니다. 이러한 차이점이 어떠한 장점과 단점이 있는지 알고 싶습니다.</div>
<li>함수의 기능 분리</li>
<div> -함수의 기능분리에 대해서 많은 고민을 하고 이번 코드를 작성 하였는데 기능분리가 부족한 부분이 있을까요?</div>
</ul>

Developer-Groo committed:
Refactor: 에러타입 제거 및 inputUserNumber함수 수정

Charl-es committed:
Refactor: 함수명 수정 및 기능분리

playGame 함수명을 startGame 함수명으로 변경.
startGame의 분기처리 로직을 새로운 함수를 정의하여 기능 분리.

Developer-Groo committed:
Feat: Mukchippa  타입 설계 및 MukchippaManager 구조체 설계

Charl-es committed:
Feat: 현재의 턴이 누구의 턴인지를 반환하는 함수 구현

Developer-Groo committed:
Feat: MukchippaManager 2개의 변수 추가

-turn을 표시하는 isTurn변수 추가
-승패를 표시하는 isWinOrLose변수 추가

Charl-es committed:
Feat: inputUserNumber 함수 추가

Developer-Groo committed:
Feat: isUserTurn 변수에 turn을 저장하는 함수 구현

Charl-es committed:
Feat: 변수 2개 선언 및 compareValues 함수 구현

컴퓨터패와 사용자패를 담는 변수 선언
컴퓨터패와 사용자패를 비교하여 승패 여부를 확인하는 함수 구현

Developer-Groo committed:
Feat: branchProcessingNumber(with userNumber:)함수 구현

숫자를 입력 받았을때 분기처리 하는 함수

Charl-es committed:
Feat: startMukchippa 함수 구현

묵찌빠 게임을 시작하는 함수 구현

mint3382 commented on RockPaperScissors/RockPaperScissors.xcodeproj/project.pbxproj:
```diff
@@ -7,8 +7,9 @@
 		C7ABA0E1254EA63000B2367D /* main.swift in Sources */ = {isa = PBXBuildFile; fileRef = C7ABA0E0254EA63000B2367D /* main.swift */; };
-		DC43FB982AA9CE1100A1B23A /* InputError.swift in Sources */ = {isa = PBXBuildFile; fileRef = DC43FB972AA9CE1100A1B23A /* InputError.swift */; };
 		DC9024D22AA6C09400D44FF8 /* RockPaperScissors.swift in Sources */ = {isa = PBXBuildFile; fileRef = DC9024D12AA6C09400D44FF8 /* RockPaperScissors.swift */; };
 		DC9024D42AA6C0A400D44FF8 /* RPSManager.swift in Sources */ = {isa = PBXBuildFile; fileRef = DC9024D32AA6C0A400D44FF8 /* RPSManager.swift */; };
```
파일명은 변경이 안되어 있습니다.🥲

Developer-Groo replied:
파일명 수정했습니다!

mint3382 commented on RockPaperScissors/RockPaperScissors/MukchippaManager.swift:
```diff
@@ -0,0 +1,81 @@
+    
+    private func printMenu(turn: String) {
+        print("[\(turn)턴] 묵(1), 찌(2), 빠(3)! <종료 : 0> :", terminator: " ")
+    }
```
매개변수로 받아서 사용하는 점 아주 좋습니다! 👍
그런데 trun이 String type이면 "사용자"와 "컴퓨터" 말고도 이 printMenu(turn: "민트")라고 어딘가에서 사용해버린다면 [민트 턴]이 될 수 있겠네요? 사용자와 컴퓨터 이 2가지 경우만 입력 받을 수 있게 만들어보는 것은 어때요?

Developer-Groo replied:
수정했습니다!

mint3382 commented on RockPaperScissors/RockPaperScissors/MukchippaManager.swift:
```diff
@@ -0,0 +1,81 @@
+    }
+    
+    private mutating func inputUserNumber() -> Int {
+        while true {
```
while true의 장단점에 대해서 설명해주세요 😼

Developer-Groo replied:
장점 : 조건이 항상 true이기 때문에 while true문 내부에서 로직을 작성하는데 자유도가 높은 것 같습니다.
단점 : 조건이 항상 true이기 때문에 while true문 내부에서 적절한 제어전송문을 사용해 주지 않으면 무한루프에 빠질수 있습니다.
장점이 단점이 될수도 있기 때문에 적절한 논리구조를 만들고 제어전송문을 상황에 맞게 사용해주어야 한다고 생각합니다.

mint3382 commented on RockPaperScissors/RockPaperScissors/MukchippaManager.swift:
```diff
@@ -0,0 +1,81 @@
+                print("잘못된 입력입니다. 다시 시도해주세요.")
+                isUserTurn = false
+                checkPlayerTurn()
+            }
```
바인딩을 하면서 동시에 게임 진행 과정 2개를 서로 구분해주니 절대 나올일 없는 값을 return 해주거나 error handling을 하지 않아도 되는 코드가 되었군요! 너무 좋아요~ 👏

mint3382 commented on RockPaperScissors/RockPaperScissors/MukchippaManager.swift:
```diff
@@ -0,0 +1,81 @@
+        }
+    }
+    
+    private mutating func checkPlayerTurn() {
```
User와 Player가 동시에 쓰이니 네이밍이 서로 헷갈리는 것 같아요. Player도 보통 게임을 하는 사용자의 의미로 쓰이기 때문에 혼동이 있어 보입니다. 다른 네이밍을 고려해보는 것은 어때요?

Developer-Groo replied:
네이밍 수정 했습니다!

mint3382 replied:
Player가 User로 바뀌었던데 그러면 사용자의 턴인지 확인만 하는 걸까요?
confirmTurn과 같은 네이밍은 어떨까요?

mint3382 commented on RockPaperScissors/RockPaperScissors/MukchippaManager.swift:
```diff
@@ -0,0 +1,81 @@
+        } else {
+            printMenu(turn: "컴퓨터")
+        }
+    }
```
기능 분리👍

mint3382 commented on RockPaperScissors/RockPaperScissors/MukchippaManager.swift:
```diff
@@ -0,0 +1,81 @@
+    
+    private mutating func compareValues() {
+        if computerCard == userCard {
+            isUserTurn ? print("사용자의 승리!") : print("컴퓨터의 승리!")
```
삼항 연산자 👍

mint3382 commented on RockPaperScissors/RockPaperScissors/MukchippaManager.swift:
```diff
@@ -0,0 +1,81 @@
+        } else if computerCard == .ppa && userCard == .muk ||
+                  computerCard == .muk && userCard == .chi ||
+                  computerCard == .chi && userCard == .ppa {
+            print("컴퓨터의 터입니다.")
```
오타가 있네요 🥲

Developer-Groo replied:
수정했습니다!

mint3382 commented on RockPaperScissors/RockPaperScissors/MukchippaManager.swift:
```diff
@@ -0,0 +1,81 @@
+        }
+    }
+    
+    private mutating func branchProcessingNumber(with userNumber: Int) {
```
이 네이밍은 함수의 역할 보다는 이 함수의 기능에 대한 설명 같네요. 
구조 내에서 어떠한 역할을 하고 있는지 고민해보고 더 알맞는 네이밍을 선사해줍시다!

Developer-Groo replied:
수정했습니다!

mint3382 commented on RockPaperScissors/RockPaperScissors/MukchippaManager.swift:
```diff
@@ -0,0 +1,81 @@
+    }
+    
+    private mutating func branchProcessingNumber(with userNumber: Int) {
+        switch userNumber {
+        case 0:
+            print("게임종료")
+            return
+        case 1...3:
+            compareValues()
+            return
+        default:
+            print("잘못된 입력입니다. 다시 시도해주세요.")
+            isUserTurn = false
+            startMukchippa()
+        }
```
switch를 알맞게 사용해서 더 깔끔한 로직이 된 것 같아요👍 
그런데 case를 처음 보았을 때 단순히 숫자 뿐이라 왜 게임이 종료되고, 진행되는지 등에 대한 설명이 부족해보입니다.

mint3382 commented on RockPaperScissors/RockPaperScissors/RPSManager.swift:
```diff
@@ -8,58 +8,66 @@
 struct RockPaperScissorsManager {
     private var computerCard: RockPaperScissors?
     private var userCard: RockPaperScissors?
+    private var mukchippa = MukchippaManager()
```
가위바위보 게임 매니저에서 왜 묵찌빠 매니저를 가지고 있나요? 가위바위보 매니저는 가위바위보만 관리해야하지 않을까요?

Developer-Groo replied:
가위바위보 게임 매니저와 묵찌빠 매니저를 따로 분리해서 수정했습니다!

mint3382 replied:
아주 좋습니다! 더불어서 main에서 if 문으로 처리를 해주고 계시던데 그러한 처리 로직이 있는 객체도 따로 있으면 좋을 것 같네요! 메인에서는 호출만 할 수 있도록이요! 😄

mint3382 commented on RockPaperScissors/RockPaperScissors/RPSManager.swift:
```diff
@@ -8,58 +8,66 @@
-        
-        return number
+    private func inputUserNumber() -> Int {
+        while true {
```
while true나 재귀 말고 어떤 방법을 쓸 수 있을까요?

Developer-Groo replied:
for 문을 생각해보긴 했지만 for문을 사용하는 경우는 반복의 횟수가 정해져 있을때 사용하는 문법이라고 알고 있기 때문에 해당 상황에서는 적절하지 않다고 판단을 했습니다. while true문과 재귀 말고 다른 적절한 방법은 생각하지 못했습니다..

mint3382 replied:
flag를 세우는 방법도 있을 것 같네요! while true와 크게 다를 것은 없지만 흐름 있게 조건을 조절해주고, 가독성에도 도움이 될 것 같습니다.

mint3382 commented on RockPaperScissors/RockPaperScissors/RPSManager.swift:
```diff
@@ -8,58 +8,66 @@
+            return
+        case 1...3:
+            compareValues(computerValue: userCard, userValue: computerCard)
+            return
```
여기서도 case 0, 1, 2,3이 어떠한 경우를 나타내는 것인지에 대한 명시가 필요해 보입니다.

Developer-Groo replied:
case 마다 실행시키는 부분에서 0인 경우에는 "게임종료"라고 print문이 작성 되어 있고 1...3 case에는 compareValues 함수를 실행시키고 있어서 case가 어떠한 경우를 나타내는지 명확하게 알 수 있다고 생각됩니다..혹시 저희가 다르게 이해한 부분이 있다면 설명주시면 감사하겠습니다!!

mint3382 replied:
case가 숫자로만 되어 있으니 밑에 "게임종료"가 있어도 제대로 된 것인지 아닌지 한 눈에 알아보기가 어렵습니다. 이번 프로젝트가 파일도 몇개 만들지 않는 작은 프로젝트라 읽고 이해할 수 있지만 커지면 커질수록 0,1,2,3과 같은 숫자만으로는 정확하게 이게 왜 이렇게 나누어서 진행했는지 알아보기가 어려울 것 같습니다. 이러한 이유 때문에도 저희는 enum을 사용하고는 합니다. 가독성을 위해서요!

mint3382 commented on RockPaperScissors/RockPaperScissors/RPSManager.swift:
```diff
@@ -8,58 +8,66 @@
-                playGame()
-            }
-        } catch InputError.wrongNumber {
+    private mutating func playingGame(with userNumber: Int) {
```
자연스럽게 읽히는 argument label의 사용 아주 좋아요🥰

mint3382 commented:
안녕하십니까, 찰스 토이!
스텝 2 진행하시느라 정말 고생하셨습니다!
많은 고민과 이전 코멘트 반영을 하려 노력하신 부분들이 보여서 너무 좋았어요🥰

## 칭찬하고 싶은 점 👍
- 구체적이고 세분화 된 커밋 내역
- 함수 기능 분리
- argument label을 이용한 가독성
- 삼항 연산자 사용
- 지난 스텝에서 if문을 사용하였다면 이번에는 switch로 바꿔보는 시도
- 스텝 별 브랜치의 분리



> ## 고민했던 점
>컴퓨터의 턴과 사용자의 턴을 구분하는 방법을 고민했습니다.
>-isUserTurn변수를 사용해서 컴퓨터의 턴과 사용자의 턴을 구분하여 담을 수 있도록 해결했습니다.
>-checkPlayerTurn함수를 사용해서 isUserTurn변수의 값에 따라 턴과 메뉴를 출력하도록 구현했습니다.

Bool 타입으로 flag를 세워준 점 아주 좋다고 생각합니다. 👍

>## 조언을 얻고 싶은 점
>### case를 나누는 단위
>-가위,바위,보와 묵,찌,빠만 case를 나눠서 구현을 했는데 다른 팀의 코드를 보니 승패여부 등 세세하게 case를 나눠서 구현을 한 것 을 확인했습니다. 이러한 차이점이 어떠한 장점과 단점이 있는지 알고 싶습니다.

enum은 코드를 짤 때 선택지를 미리 주어서 해당 선택지 외에 다른 선택을 하지 못하게 하기 때문에 기능적으로 잘못된 선택지가 들어갈 일을 줄일 수 있습니다. 또한 목록으로 나열할 수 있는 경우 구분의 가독성을 위해 사용합니다. 때문에 선택해야하는 선택지의 경우, 특히 그것을 매개 변수로 서로 전달받을 경우에는 enum을 사용하는 것이 기능적으로도, 가독성 면에서도 좋을 것 같습니다. 다만 enum은 모든 case가 동일한 타입의 rawValue를 가져야 하고 그것도 한 가지 값 밖에 사용할 수 없다는 한계가 있었는데 이를 해결하기 위해 associated type이라고 보완할 수 있는 방법이 등장했습니다. 때문에 위에 적어 놓은 것처럼 선택지의 한계를 주고 싶을 때는 enum을 사용하는 편이 좋다고 생각합니다. 

>### 함수의 기능 분리
>-함수의 기능분리에 대해서 많은 고민을 하고 이번 코드를 작성 하였는데 기능분리가 부족한 부분이 있을까요?

이번 스텝 2에서 기능 분리에 대해 많은 고민을 하신 것 같고, 그런 만큼 함수들의 기능 분리가 잘 되어 있었습니다. 🥰 

다만 이제 함수의 기능 분리를 넘어서 객체들의 관계에 대한 고민을 해보시면 좋을 것 같습니다. 가위바위보와 묵찌빠는 두 개 다 게임인데 어떤 관계가 있을까요? 가위바위보만 하고 싶어서 가위바위보 매니저만 생성했을 때 불필요한 묵찌빠를 알게 되진 않을까요? 깊은 공부를 하고 싶으시다면 protocol에 대해 찾아보시는 것도 추천드립니다. 😆 

고생 많으셨고, 응원합니다! 파이팅~

mint3382 commented on RockPaperScissors/RockPaperScissors/MukchippaManager.swift:
```diff
@@ -0,0 +1,81 @@
+//
+
+struct MukchippaManager {
+    var isUserTurn = true
```
Bool 타입에 맞는 네이밍👍

mint3382 commented on RockPaperScissors/RockPaperScissors/MukchippaManager.swift:
```diff
@@ -0,0 +1,81 @@
+struct MukchippaManager {
+    var isUserTurn = true
+    private var computerCard: Mukchippa?
+    private var userCard: Mukchippa?
```
isUserTurn은 타입 선언 없이 바로 값을 가지고 있고 computerCard와 UserCard는 타입만 가지고 있는 것이 컨벤션이 두 개가 서로 달라 보입니다. 통일해주시면 좋겠습니다!

Developer-Groo replied:
수정했습니다!

mint3382 commented on RockPaperScissors/RockPaperScissors/RPSManager.swift:
```diff
@@ -8,58 +8,66 @@
 struct RockPaperScissorsManager {
     private var computerCard: RockPaperScissors?
     private var userCard: RockPaperScissors?
+    private var mukchippa = MukchippaManager()
```
여기도 computerCard & userCard와 mukchippa 변수의 컨벤션이 달라 보입니다. 타입 선언을 통일해주세요!

Developer-Groo replied:
수정했습니다!

Charl-es committed:
Refactor: RPSManager 파일명 수정, chekcUserTurn 로직 수정

Developer-Groo committed:
Refactor: 타입 컨벤션이 통일 되도록 수정

Charl-es committed:
branchProcessingNumber 함수명 변경. 오탈자 수정

Developer-Groo committed:
Refactor: RockPaperScissorsManager와 MukchippaManager를 분리

RockPaperScissorsManager에 있던 MukchippaManager를 분리하여 main.swift파일에서 분기 처리를 하도록 수정

Charl-es committed:
doc: Update README.md
